crates/subspace-service/src/sync_from_dsn/fast_sync.rs

```diff
@@ -0,0 +1,587 @@
+        node: &'a Node,
+        piece_getter: &'a PG,
+        client: Arc<Client>,
+        import_queue_service: Arc<Mutex<Box<IQS>>>,
+        network_service: Arc<NetworkService<Block, <Block as BlockT>::Hash>>,
+        fast_sync_state: Arc<parking_lot::Mutex<Option<NumberFor<Block>>>>,
+        sync_service: Arc<SyncingService<Block>>,
+    ) -> Self {
+        Self {
+            segment_headers_store,
+            node,
+            piece_getter,
+            client,
+            import_queue_service,
+            network_service,
+            fast_sync_state,
+            sync_service,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Sync algorithm:
+    /// - 1. download two last segments,
+    /// - 2. add the last block the second last segment (as raw - without checks and execution),
+    /// - 3. download the state for the first block of the last segment,
+    ///     - add the first block of the last segment as raw,
+    ///     - download state for it,
+    /// - 4. add blocks with execution from the last segment,
+    ///     - notify Substrate sync by importing block with using BlockOrigin::NetworkBroadcast,
+    ///     - clear the block gap to prevent Substrate sync to download blocks from the start.
+    ///     - restart the Substrate sync with SyncMode::Full to enable it downloading full blocks.
+    ///  Note: retry for fast sync will degrade reputation if we have already announced blocks.
+    pub(crate) async fn sync(&self) -> Result<FastSyncResult<Block>, Error> {
+        debug!("Starting fast sync...");
+
+        // 1. Download the last two segments.
+        let mut reconstructor = Reconstructor::new().map_err(|error| error.to_string())?;
+        let segment_header_downloader = SegmentHeaderDownloader::new(self.node);
+
+        if let Err(error) = self
+            .download_segment_headers(&segment_header_downloader)
+            .await
+        {
+            error!(?error, "Failed to download segment headers.");
+            return Err(error);
+        };
+
+        let Some(last_segment_index) = self.segment_headers_store.max_segment_index() else {
+            return Err(Error::Other("Can't get last segment index.".into()));
+        };
+
+        // Skip the fast sync if we lack the minimum required segment number
+        if last_segment_index <= SegmentIndex::ONE {
+            return Ok(FastSyncResult::skipped());
+        }
+
+        let last_segment_header = self
+            .segment_headers_store
+            .get_segment_header(last_segment_index)
+            .expect("We get segment index from the same storage. It should be present.");
+
+        let second_last_segment_index = last_segment_header.segment_index() - 1.into();
+
+        let second_last_segment_blocks = download_and_reconstruct_blocks(
+            second_last_segment_index,
+            self.piece_getter,
+            &mut reconstructor,
+        )
+        .await?;
```

Comment:
This is not how algorithm should work according to the spec though. You're supposed to check the segment header and only download second last semgnet header in case it is actually necessary, which is most of the time, but not always.

Reply:
The only thing that needs to be done is an `if` branch. If the last block in second last segment was included partially then we download the segment and feed it into reconstructor, otherwise nothing needs to be downloaded or fed into reconstructor and we can proceed with the last segment only. This is a very simple check and I don't understand why we should intentionally not implement the spec and describe a different version in the function signature.

And such things are already tested in the archiver, you can easily logically confirm how it will behave without actually running the code.

I do not think I understand what the drawbacks there are in following the spec.

Reply:
The reason is that we won't test this code and it can fail even if seems logical which happened multiple times with other simple changes. It's likely possible to emulate this case when we have a complete last archived block. I checked the current blockchain - we don't have it currently. I added a TODO with a reference to the spec difference.

Reply:
I'm sorry but it makes no sense to me. If we are so afraid to make changes we'll not make progress at any meaningful rate. Writing code that is intentionally wrong against the spec and even strictly speaking broken in general case makes even less sense to me.